scipy/stats/_continuous_distns.py

```diff
@@ -2923,6 +2923,64 @@ def fit(self, data, *args, **kwds):
 gamma = gamma_gen(a=0.0, name='gamma')
 
 
+class gamma_gompertz_gen(rv_continuous):
+    u""" A Gamma/Gompertz continuous random variable
+
+    %(before_notes)s
+
+    Notes
+    -----
+    The probability density function for `gamma_gompertz` is:
+
+        gamma_gompertz.pdf(x, c, beta) = c * exp(x) * (beta**c) / ((beta - 1 + exp(x))**(c + 1))
+
+    for ``x >= 0``, ``c > 0``, ``beta > 0``.
+
+    `gamma_gompertz` takes ``c`` and ``beta`` as shape parameters.
+
+    %(after_notes)s
+
+    References
+    ----------
+    .. [1] https://en.wikipedia.org/wiki/Gamma/Gompertz_distribution
+
+    %(example)s
+    """
+
+    def _logsf(self, x, c, beta):
+        em1 = sc.expm1(x)
+        # np.where is used for dealing with the case that x is too large
+        # so that sc.expm1(x) becomes infinity.
+        # Since x is very large in that case, -sc.log1p(em1 / beta)
+        # approximately equals to -sc.log(np.exp(x) / beta) = np.log(beta) - x
+        lw = _lazywhere(np.isfinite(em1), [em1, x, beta],
+                        lambda em1_, x_, beta_: -sc.log1p(em1_ / beta_),
+                        f2 = lambda em1_, x_, beta_: np.log(beta_) - x_)
+        return c * lw
+
+    def _sf(self, x, c, beta):
+        return np.exp(self._logsf(x, c, beta))
+
+    def _cdf(self, x, c, beta):
+        return 1. - (beta ** c) / (beta + sc.expm1(x))
+
+    def _pdf(self, x, c, beta):
+        return np.exp(self._logpdf(x, c, beta))
+
+    def _logpdf(self, x, c, beta):
+        em1 = sc.expm1(x)
+        lw = _lazywhere(np.isfinite(em1), [em1, x, beta],
```

Comment:
I think the best way is to write beta + e^x - 1 = e^x * (1 + e^-x (beta - 1)). Taking logs, you get x + log1p(e^-x (beta - 1)). This should work well for large x and no need to distinguish cases

Reply:
There are multiple ways to write this expression, each of which works well in some region.  I don't know if there is a single one which will always be sufficient. (E.g. `x + log1p(e^-x (beta - 1))` may lose precision if beta and x are both close to 0.)  But my original suggestion of `np.where(np.isfinite(sc.expm1(x)), np.log(beta + sc.expm1(x)), x)` is clearly inadequate as it is not sufficient to check whether `e^x-1` is finite since `beta` still needs together added to it.

Reply:
Thanks for the discussions.

**My Opinion:**
1. Parameters `x`, `c`, `beta` are usually neither too large nor too small. `c` and `beta` are usually 0.1~10. `x` can be hundreds, but `x=1000` is not rare either.
2. My current workaround aims to resolve the case when `x` is not too large, but `exp(x)` is infinity. For example, when `x=1000`, `exp(x)=+inf` and `sc.expm1(x)=+inf`,
3. `beta` can indeed be large if user intentionally feed a large `beta`. However, when `beta` is not `inf`, it does not prevent the most of calculation from generating a reasonable answer.

Therefore, I think my implementation is acceptable.

**Example 1:** (small `beta`)
Let us calculate `logpdf(x=1000, c=1, beta=2) = 1000+log(2)-2*log1p(exp(1000)) \approx -999.307 `.
- The normal branch of my implementation: `lw=np.log(beta_ + em1_)=log(2+inf)=inf`, `pdf=np.log(c) + x + c * np.log(beta) - (c + 1.) * lw=log(1)+1000+1*log(2)-2*+inf=-inf` (Incorrect)
- The alternative branch of my implementation: `lw=x_=1000`, `pdf=np.log(c) + x + c * np.log(beta) - (c + 1.) * lw=log(1)+1000+1*log(2)-2*1000=log(2)-1000 \approx -999.307` (Correct)
- The implementation of `e^x * (1 + e^-x (beta - 1))`, `e^x * (1 + e^-x (beta - 1))=+inf*(1+ 0 *(2-1))=+inf`, `pdf=c*exp(x)*(beta**c)/+inf=1*+inf*2/+inf=nan` (Incorrect)

**Example 2:** (large `beta`)
Let us calculate `logpdf(x=1000, c=1, beta=1000) = 1000+log(1000)-2*log(999+exp(1000)) \approx -993.092 `.
- The normal branch of my implementation: `lw=np.log(beta_ + em1_)=log(1000+inf)=inf`, `pdf=np.log(c) + x + c * np.log(beta) - (c + 1.) * lw=log(1)+1000+1*log(2)+log(1000)-2*+inf=-inf` (Incorrect)
- The alternative branch of my implementation: `lw=x_=1000`, `pdf=np.log(c) + x + c * np.log(beta) - (c + 1.) * lw=log(1)+1000+1*log(1000)-2*1000=log(1000)-1000 \approx -993.092` (Correct)
- The implementation of `e^x * (1 + e^-x (beta - 1))`: `e^x * (1 + e^-x (beta - 1))=+inf*(1+ 0 *(2-1))=+inf`, `pdf=c*exp(x)*(beta**c)/+inf=1*+inf*1000/+inf=nan` (Incorrect)

Reply:
One still needs to take the log after rewriting it:

`log(e^x * (1 + e^-x (beta - 1))) = x + log1p(...)`

> I don't know if there is a single one which will always be sufficient. (E.g. x + log1p(e^-x (beta - 1)) may lose precision if beta and x are both close to 0.)

one could potentially use the current implementation for small x and the x + log1p(...) for large x (I haven't thought about it)

Reply:
I still think this part of the implementation needs to be fixed

Reply:
@chrisb83

Thanks for your attention. I do not think small `x` can lead to numerical issues.

**Example** (small `x`):
Let us calculate `logpdf(x=1e-8, c, beta=2)`. In theory, `log(beta+exp(x)-1)=log1p(exp(1e-8))` approx `log(2)=0.6931`.

- The normal branch of my implementation: `log(beta+expm1(1e-8))=log(2+1e-8)` approx `log(2)=0.6931` (Correct)
- The implementation of `x+log1p(exp(-x)(beta-1))=(1e-8)+log1p(exp(-1e-8))` approx `1e-8+log(2)=0.6931` (Correct)

Reply:
logpdf:
```
bem1 = beta + sc.expm1(x)
lw = _lazywhere(np.isfinite(bem1), [bem1, x],
                          lambda bem1_, x_: np.log(bem1_), # this part is OK
                          f2 = lambda bem1_, x_: x_) # this part does not work
return np.log(c) + x + c * np.log(beta) - (c + 1.) * lw 
```
the second part does not work if both x and beta are large, see explanation above. I think you can rewrite this part as x + log1p(e^-x (beta - 1))

logsf might have a similar issue

Reply:
@chrisb83 The **Example 2** above has both large `beta` and large `x`. Could you please give an example where my implementation does not work? Thanks in advance!

Reply:
It is a constructed example, I am not saying it is a typical use case but it show that one has to be careful with the approximation:

```
x = 710 # on my system, np.expm1(x) = inf, whereas it is finite for x - 1
beta = np.expm1(709)
```
The term `lw` would be 710 in your code.
But `x + np.log1p(np.exp(-x) * (beta - 1)) = 710.31...` and depending on `c`, the end result can be completely off. For example, with `c=5`, your code gives logpdf = -3.39... instead of -5.27...

Reply:
@chrisb83   Thanks for your example. It helps.
Accordingly, I revised the implementation of `logpdf` and `logsf`.